gramps/plugins/db/dbapi/dbapi.py

```diff
@@ -1699,3 +1732,211 @@ def bulk_get_families(self, handles):
+            # Fallback to individual commits
+            for person in persons:
+                self._commit_person(person, trans)
+                
+        # Apply auxiliary updates (COMPLETING THE IMPLEMENTATION)
+        from gramps.gen.lib import Person
+        
+        for person in persons:
+            old_data = old_data_map.get(person.handle)
+            
+            if old_data:
+                # Deserialize old person for comparison
+                old_person = self.serializer.string_to_object(old_data, Person)
+                
+                # Update gender statistics if necessary
+                if (old_person.gender != person.gender or 
+                    old_person.primary_name.first_name != person.primary_name.first_name):
+                    self.genderStats.uncount_person(old_person)
+                    self.genderStats.count_person(person)
+                    
+                # Update surname list if necessary
+                if self._order_by_person_key(person) != self._order_by_person_key(old_person):
+                    self.remove_from_surname_list(old_person)
+                    self.add_to_surname_list(person, trans.batch)
+            else:
+                # New person - add to auxiliary structures
+                self.genderStats.count_person(person)
+                self.add_to_surname_list(person, trans.batch)
+            
+            # Type registry updates (same as commit_person)
+            self.individual_attributes.update(
+                [str(attr.type) for attr in person.attribute_list 
+                 if attr.type.is_custom() and str(attr.type)]
+            )
+            
+            self.event_role_names.update(
+                [str(eref.role) for eref in person.event_ref_list 
+                 if eref.role.is_custom()]
+            )
+            
+            self.name_types.update(
+                [str(name.type) for name in ([person.primary_name] + person.alternate_names)
+                 if name.type.is_custom()]
+            )
+            
+            all_surn = []
+            all_surn += person.primary_name.get_surname_list()
+            for asurname in person.alternate_names:
+                all_surn += asurname.get_surname_list()
+            self.origin_types.update(
+                [str(surn.origintype) for surn in all_surn 
+                 if surn.origintype.is_custom()]
+            )
+            
+            self.url_types.update(
+                [str(url.type) for url in person.urls 
+                 if url.type.is_custom()]
+            )
+            
+            attr_list = []
+            for mref in person.media_list:
+                attr_list += [str(attr.type) for attr in mref.attribute_list 
+                             if attr.type.is_custom() and str(attr.type)]
+            self.media_attributes.update(attr_list)
+            
+            # Emit signal for GUI updates
+            self.emit('person-add', ([person.handle],))
```

Comment:
by emitting within the for loop, a person-add signal is generated whilst the DB is in an inconsistent state; the Person records have been fully updated but we have not yet made all of the corresponding updates to generstats, surname lists, individual attributes etc.
Is it better to complete all data updates and then have a second loop to emit the signals? That way the data is fully consistent.